src/DurableTask.Core/Entities/EntityId.cs

```diff
@@ -0,0 +1,111 @@
+//  You may obtain a copy of the License at
+//  http://www.apache.org/licenses/LICENSE-2.0
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//  ----------------------------------------------------------------------------------
+namespace DurableTask.Core.Entities
+{
+    using System;
+    using System.Runtime.Serialization;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// A unique identifier for an entity, consisting of entity name and entity key.
+    /// </summary>
+    [DataContract]
+    public readonly struct EntityId : IEquatable<EntityId>, IComparable
+    {
+        /// <summary>
+        /// Create an entity id for an entity.
+        /// </summary>
+        /// <param name="entityName">The name of this class of entities.</param>
+        /// <param name="entityKey">The entity key.</param>
+        public EntityId(string entityName, string entityKey)
+        {
+            if (string.IsNullOrEmpty(entityName))
+            {
+                throw new ArgumentNullException(nameof(entityName), "Invalid entity id: entity name must not be a null or empty string.");
+            }
+
+            this.Name = entityName;
+            this.Key = entityKey ?? throw new ArgumentNullException(nameof(entityKey), "Invalid entity id: entity key must not be null.");
+        }
+
+        /// <summary>
+        /// The name for this class of entities.
+        /// </summary>
+        [DataMember(Name = "name", IsRequired = true)]
+        public readonly string Name;
+
+        /// <summary>
+        /// The entity key. Uniquely identifies an entity among all entities of the same name.
+        /// </summary>
+        [DataMember(Name = "key", IsRequired = true)]
+        public readonly string Key;
+
+        /// <inheritdoc/>
+        public override string ToString()
+        {
+            return $"@{this.Name}@{this.Key}";
+        }
+
+        internal static string GetSchedulerIdPrefixFromEntityName(string entityName)
+        {
+            return $"@{entityName}@";
+        }
+
+        /// <summary>
+        /// Returns the entity ID for a given instance ID.
+        /// </summary>
+        /// <param name="instanceId">The instance ID.</param>
+        /// <returns>the corresponding entity ID.</returns>
+        public static EntityId FromString(string instanceId)
+        {
+            if (instanceId == null)
+            {
+                throw new ArgumentNullException(nameof(instanceId));
+            }
+            var pos = instanceId.IndexOf('@', 1);
+            if (pos <= 0)
+            {
+                throw new ArgumentException("instanceId is not a valid entityId", nameof(instanceId));
```

Comment:
Consider including the actual instance ID value in the exception to make it easier to debug if we ever encounter this.

```suggestion
                throw new ArgumentException($"Instance ID '{instanceId}' is not a valid entity ID.", nameof(instanceId));
```

Reply:
Yes, makes sense. I will also add a check that the first character should be '@'.